conans/client/tools/oss.py

```diff
@@ -428,6 +428,10 @@ def cross_building(settings, self_os=None, self_arch=None, skip_x64_x86=False):
     if host_os is not None and (build_os != host_os):
         return True
     if host_arch is not None and (build_arch != host_arch):
+        if host_arch == 'ppc32' and build_arch == 'ppc64':
+            return False
```

Comment:
I am not so sure about this change. This is ok for the public usage, that is done in recipes, for example to be able to execute some executable:

```python
def test(self):
    if not tools.cross_building(self.settings):
          self.run("myexe")
```

But this method is used also internally, for example in the ``AutotoolsBuildEnvironment()``, to define the gnu triplet:
```python
if not cross_building(self._conanfile.settings, os_detected, arch_detected):
      return False, False, target
```
It seems that this will not set the necessary triplet to build for 32 bits in those platforms. 

Even the case for ``x86`` "not" cross building, is an opt-in, it requires to pass ``skip_x64_x86=True`` to the ``tools.cross_building()`` method.

Reply:
@memsharded I'm struggling to fully understand the use case you point out, in terms of how a user sets `skip_x64_x86`.   Could you elaborate on that a bit?

I think the other thing to point out is that this change may not be 100% perfect, but it should allow for building 32-bit on SPARC and Power architectures natively, which currently it will not do.  If it breaks cross builds on SPARC and Power (not target, but host) that would be a much less common use case IMO.
The other difference is that IIRC, Autoconf will NOT generate a different host identifier for building 32-bit architectures.  Autoconf doesn't attempt to guess if we're cross building, it won't assume a cross unless the target architecture is *explicitly* specified.

Reply:
Yes, the current implemented behavior is:
- x86 will be considered cross-building from x86_64, unless the user explicitly tells ``tools.cross_building(...., skip_x64_x86=True)``. It is an opt-in, not the default behavior
- This is done to guarantee that it will not break other users that still want to consider, or rely on the previous behavior that such helper returned ``True`` for this case
- The ``AutotoolsBuildEnvironment`` creates a "gnu-triplet" that is passed to the underlying autoconf/make system. The helper will use ``tools.cross_building`` to define a "gnu-triplet" that can look like ``x86-xxxx``. It will be passed to autotools while calling them as an argument ``--build=xxx``, ``--host=xxx`` and/or ``--target=xxx```. 
- If ``tools.cross_building()`` starts to return False for this case (or other), then the triplet might be empty, and that might break other users as well. When it was added as an opt-in with ``skip_x64_x86``, then it was guaranteed that this wouldn't break.

Please let me know if this clarify a bit my concerns.

Reply:
I see that for my 32-bit Linux build on x86_64 Linux example, the `cross_building` method is returning `True` when `skip_x64_x86` is `False`.
I'm guessing this must be special-cased somewhere else to ignore it, or the same problems I see on AIX & Solaris would be present - in my case the conan-icu project attempts to install the icu binaries on the build machine before it attempts to "cross" build.

The use in `AutotoolsBuildEnviornment()` seems correct actually - if you can build & run on the same machine Autotools would correctly build without setting the GNU triplet explicitly.  Autoconf won't assume a cross build unless the triplet is explicitly supplied, and we shouldn't do that for 32-bit binaries on a 64-bit machine (at least not the ones I'm familiar with).

Reply:
> I see that for my 32-bit Linux build on x86_64 Linux example, the cross_building method is returning True when skip_x64_x86 is False.

Yes, that is the logic:
- When building to 32 bits in 64 bits platform.
      - With ``skip_x64_x86=False`` ``tools.cross_build()=> True``
      - With ``skip_x64_x86=True`` ``tools.cross_build()=> False``


> The use in AutotoolsBuildEnvironment() seems correct actually - if you can build & run on the same machine Autotools would correctly build without setting the GNU triplet explicitly. Autoconf won't assume a cross build unless the triplet is explicitly supplied, and we shouldn't do that for 32-bit binaries on a 64-bit machine (at least not the ones I'm familiar with).

Ok, so for that case, it could be possible that the ``-m32`` flag (also provided by the ``AutotoolsBuildEnvironment()`` would be enough, and no need to specify the gnu triplet. But the fact is that when ``-s arch=x86`` is specified, the triplet that is used is:

```
build = x86_64-linux-gnu
host = x86_64-linux-gnu
```
These are passed to the ``configure`` command.

This is the current behavior, not sure if this clarifies a bit.

Reply:
Thanks @memsharded that clears things up.
I think it's reasonable to change the behavior of a public interface when, as in this case, the behavior is the bug.
Where it mattered, I have already worked around this problem in specific packages, which I'm sure everyone who had this problem with 32-bit builds on amd64 has already done.  Now that I understand it fully, my first choice would be to ignore `skip_x64_x86`, my second choice to extend that flag to mean any 32-bit build on a 64-bit machine.